src/sentry/eventstream/kafka/backend.py

```diff
@@ -148,6 +148,18 @@ def _send(
             # flush() is a convenience method that calls poll() until len() is zero
             self.producer.flush()
 
+    @staticmethod
+    def _key(headers: Optional[Mapping[str, str]], project_id: str) -> Optional[str]:
+        if "transaction_forwarder" in headers:
+            message_type = "error" if headers["transaction_forwarder"] == "0" else "transaction"
```

Comment:
I think this is going to be very error prone. 
We are coupling the code that does the routing of the message to the code that decides the partitioning. While they depend on the same condition, there is no reason for these to depend on each other. 
Basically if the code that sets the header changes (which is in a totally different place) this code would likely break for no reason and in a way that will be difficult to track down.
This will be made worse by the fact that this code is not really easy to test.

If you want to reuse the same logic that we use to fill in the `transaction_forwarder` field, please take that logic out of the ` _get_headers_for_insert` method of `KafkaEventStream` and reuse it here.

Reply:
Right, though I think the reason I did it this way is that `_get_headers_for_insert` relies on the event object which `_send` doesn't have access to. Do you think event or group_id or some kind of meta structure should be a new argument to `_send()`, interface defined here: 

https://github.com/getsentry/sentry/blob/b097d4036a813a0ed5e570228b5674bb3cff8e51/src/sentry/eventstream/snuba.py#L302-L310

It does look like `group_id` is in `extra_data`:

https://github.com/getsentry/sentry/blob/b097d4036a813a0ed5e570228b5674bb3cff8e51/src/sentry/eventstream/snuba.py#L126-L131

but depending on that seems similarly brittle

Reply:
looking at this again, how would you feel about `partition_key: Optional[str]` being an argument for `_send` or perhaps `assign_partition_randomly : bool = true` and then I can make a change in this call: 

https://github.com/getsentry/sentry/blob/b097d4036a813a0ed5e570228b5674bb3cff8e51/src/sentry/eventstream/snuba.py#L126-L152